R/utils.R

```diff
@@ -0,0 +1,93 @@
+#' @return Returns \code{NULL} invisibly.
+warnMissingVars <- function(data, vars) {
+  available <- vars %in% levels(data$variable)
+  missingVars <- vars[!available]
+  if (length(missingVars) > 0)
+    warning("Variables not found: ", paste(missingVars, collapse = ", "),
+            call. = FALSE)
+  return(invisible(NULL))
+}
+
+#' Check Whether Global Options Are Provided
+#'
+#' The function checks whether a variable of the name \code{optNames} is
+#' available. If not a warning message is produced, indicating how this variable
+#' can be set using \code{options()}.
+#'
+#' @param optNames A character vector. The names of the variables.
+#' @param envir The environment where to look for the variables.
+#' @return Returns \code{NULL} invisibly.
+checkGlobalOptionsProvided <- function(optNames, envir=rlang::caller_env()) {
+  for (on in optNames) {
+    if (is.null(get0(on, envir = envir))) {
+      stop(
+        on, " must be provided, ",
+        "either as a function argument or ",
+        "as a global option via options(mip.", on, "=<value>).",
+        call. = FALSE)
+    }
+  }
+  return(invisible(NULL))
+}
+
+#' Get Legend of a ggplot.
+#'
+#' The function extracts the legend of a ggplot object.
+#'
+#' @param plt A ggplot object.
+#' @return The legend of \code{plt} or \code{NULL} if no legend was found.
+getLegend <- function(plt) {
+  tmp <- ggplot_gtable(ggplot_build(plt))
+  legIdx <- which(vapply(tmp$grobs, function(x) x$name, "") == "guide-box")
+  if (length(legIdx) == 0) return(NULL)
+  tmp$grobs[[legIdx[1]]]
+}
+#' Calculate Ratios for Quitte Objects.
+#'
+#' Changes the value of variables given in \code{numerators} by dividing by
+#' \code{denominator} and multiplying \code{conversionFactor}. Sets \code{unit}
+#' of these variables to \code{newUnit}.
+#'
+#' @param data A quitte object.
+#' @param numerators A character vector. Entries in the variable column of data.
+#' @param denominator A single string. An entry in the variable column of data.
+#' @param newUnit A single string.
+#' @param conversionFactor A single numerical value.
+#' @return A quitte object with changed values and new unit. Unused levels are dropped.
+#' @importFrom rlang .data .env
+calculateRatio <- function(
+  data, numerators, denominator, newUnit = "1", conversionFactor = 1
+) {
+  data %>%
+    filter(.data$variable == .env$denominator) %>%
+    rename(denom_value = .data$value) %>%
+    select(.data$model, .data$scenario, .data$region, .data$period, .data$denom_value) ->
+    denom
+  data %>%
+    filter(.data$variable %in% .env$numerators) %>%
+    left_join(denom) %>%
+    mutate(value = .data$value / .data$denom_value * .env$conversionFactor,
+           unit = factor(.env$newUnit)) %>%
+    droplevels() ->
+    res
+  return(res)
+}
+
+#' Longest Common Prefix
+#'
+#' @param x A character vector.
+#' @return A single string. The longest common prefix of \code{x}.
+longestCommonPrefix <- function(x) {
+  for (i in 1:min(nchar(x))) {
```

Comment:
If `x = ""` this will be `c(1, 0)`
```suggestion
  for (i in seq_len(min(nchar(x)))) {
```

Reply:
True! Thanks for spotting that!!